src/index.ts

```diff
@@ -1,220 +1,142 @@
+    crypto: Crypto.Implementation
+    wallet: Wallet.Implementation
+  }): Promise<Manners.Implementation> {
+    return BaseManners.implementation(
+      settings.wallet,
+      { configuration: settings }
+    )
+  }
 }
 
 
 
 // 🚀
 
 
-export async function app(options?: AppOptions): Promise<AppState> {
-  let fs
+export type Options = {
+  onAccountChange?: (program: Program) => unknown
+  onDisconnect?: () => unknown
 
-  options = options || {}
-
-  const { resetWnfs = false, useWnfs = true } = options
+  wallet?: Wallet.Implementation
+}
 
-  setImplementations(USE_WALLET_AUTH_IMPLEMENTATION)
 
-  // Check if browser is supported
-  if (hasProp(self, "isSecureContext") && self.isSecureContext === false) throw InitialisationError.InsecureContext
-  if (await isSupported() === false) throw InitialisationError.UnsupportedBrowser
+/**
+ * 🚀 Build a webnative-walletauth program.
+ *
+ * Contrary to a regular webnative program,
+ * this'll create an account for you automatically.
+ * It also does session management for you.
+ *
+ * You may pass in a custom `wallet` implementation
+ * if you'd like to use another wallet than the built-in
+ * Ethereum one.
+ *
+ * See [webnative's `program`](https://webnative.fission.app/functions/program-1.html) function documentation for more info.
+ */
+export async function program(settings: Options & Partial<Components> & Configuration): Promise<Program> {
+  const wallet = settings.wallet || EthereumWallet.implementation
+  const walletCrypto = await components.crypto({ ...settings, wallet })
+
+  const defaultCrypto = await Webnative.defaultCryptoComponent(settings)
+  const manners = await components.manners({ ...settings, wallet, crypto: defaultCrypto })
+
+  // Create Webnative Program
+  const webnativeProgram = await Webnative.program({
+    ...settings,
+    crypto: defaultCrypto,
+    manners,
+  })
 
-  // Authenticate & create user if necessary
+  // Destroy existing session if wallet account changed
+  const authStrategy = webnativeProgram.auth
   const username = await wallet.username()
-  const isNewUser = await isUsernameAvailable(username) === true
+  const isNewUser = await authStrategy.isUsernameAvailable(username)
 
-  let authedUsername = await authenticatedUsername()
-  if (resetWnfs || isNewUser || username !== authedUsername) {
-    await leave({ withoutRedirect: true })
-    authedUsername = null
+  let session = webnativeProgram.session
+  if (session && (isNewUser || username !== session.username)) {
+    await session.destroy()
+    session = null
   }
 
   // Initialise wallet
+  // > Destroy existing session when account changes or disconnects.
+  // > Will create a new Program on account change.
   await wallet.init({
-    onAccountChange: () => leave({ withoutRedirect: true })
-      .then(() => app(options))
-      .then(a => options?.onAccountChange ? options.onAccountChange(a) : a),
+    onAccountChange: async () => (await authStrategy.session())?.destroy()
```

Comment:
I find this and `onDisconnect` a bit hard to reason about at a glance 🤔  I think since there are a lot of moving pieces in these functions, it may help to be a bit more verbose and also add some error handling in case developers find themselves in situations where the function isn't getting invoked(like the walletauth bug yesterday).

Maybe something like this?
```
onAccountChange: async () => {
  try {
    // Destroy the user's current session when their account changes
    const session = await authStrategy.session()
    await session?.destroy()

    const p = await program(settings)
    
    // If the user has passed in a callback function, use it
    if (settings?.onAccountChange instanceof Function) {
      return settings.onAccountChange(p)
    }
    
    // Otherwise, return the program
    return p
  } catch (error) {
    console.error(error)
  }
}
```